### PR TITLE
Fixes in bhmm and some organization of header files for c++ modules + misc

### PR DESCRIPTION
**Pull request by clonker**
- organized commonly used functionality of modules in a separate folder which is included by all modules
- Quantity statistics are now a global utility
- bayesian hmm updates / fixes:
  - fixed a bug in sampling path estimation, the random generator always produced the same path
  - sampler is invoked on thin hmm data wrapper
  - the three steps of sampling (update hidden state trajs, update emission probabilities, sample hidden transition matrix) are now stateless static methods
- collecting observations in state for trajectories is now utility method and can be invoked without hmm instance
- transition count estimator is now also transformer (transforms data to count model)

### CODE REVIEW

File: sktime/base.py [hunk 1]
```diff
@@ -16,7 +17,7 @@ def __repr__(self):
             name=name, params=pprint_sklearn(self.get_params(), offset=len(name), )
         )
 
-    def get_params(self):
+    def get_params(self, deep=False):
```
Comment (brookehus) [inline sktime/base.py:20]: sorry if i'm being an idiot here but where is the part where `deep` is incorporated into what the function does?

File: sktime/markov/hmm/_bindings/include/OutputModelUtils.h [hunk 2]
```diff
@@ -49,7 +49,7 @@ np_array<dtype> generateObservationTrajectory(const np_array<State> &hiddenState
 
     #pragma omp parallel
     {
-        std::default_random_engine generator(clock() + std::hash<std::thread::id>()(std::this_thread::get_id()));
+        // std::default_random_engine generator(clock() + std::hash<std::thread::id>()(std::this_thread::get_id()));
```
Comment (brookehus) [inline sktime/markov/hmm/_bindings/include/OutputModelUtils.h:52]: why keep as opposed to delete?

File: sktime/base.py (hunk 1 above)
Comment (clonker) [reply sktime/base.py:20]: nothing wrong with that question, it is just to be API compliant with scikit-learn. Otherwise (if used in conjunction with scikit-learn methods) ducktyping will fail.

File: sktime/markov/hmm/output_model.py [hunk 3]
```diff
@@ -286,6 +289,17 @@ def sample(self, observations_per_state: List[np.ndarray]) -> None:
...
+        # N, M = self._output_probabilities.shape  # nstates, nsymbols
+        # for i, obs_by_state in enumerate(observations_per_state):
+        #     # count symbols found in data
+        #     count = np.bincount(obs_by_state, minlength=M).astype(float)
+        #     # sample dirichlet distribution
+        #     count += self.prior[i]
+        #     positive = count > 0
+        #     # if counts at all: can't sample, so leave output probabilities as they are.
+        #     self._output_probabilities[i, positive] = dirichlet(count[positive])
+        # self._output_probabilities /= np.sum(self._output_probabilities, axis=1)[:, None]
```
Comment (brookehus) [inline sktime/markov/hmm/output_model.py:302]: what's this?

File: sktime/base.py (hunk 1 above)
Comment (brookehus) [reply sktime/base.py:20]: ok, so it's correct that there's nothing here that uses any kind of `deep` parameter?

File: sktime/markov/msm/markov_state_model.py [hunk 4]
```diff
@@ -225,7 +225,11 @@ def stationary_distribution(self):
     def update_stationary_distribution(self, value: np.ndarray):
         r""" Explicitly sets the stationary distribution, re-normalizes """
         if value is not None:
-            self._stationary_distribution = np.copy(value) / np.sum(value)
+            sum = np.sum(value)
+            if sum != 0.:
```
Comment (brookehus) [inline sktime/markov/msm/markov_state_model.py:229]: out of curiosity do we ever expect this case?

File: tests/markov/hmm/test_output_models.py [hunk 5]
```diff
@@ -80,7 +80,9 @@ def test_sample(self):
             np.array([1] * 30000 + [2] * 70000)  # state 1
         ]
         m.sample(obs_per_state)
-        np.testing.assert_array_almost_equal(m.output_probabilities, np.array([[.5, .5, 0.], [0, .3, .7]]), decimal=2)
+        # the output probabilities of the unpopulated states are left as-is (can't sample), hence we compare against
+        # [[.5, .5, .1], [.1, .3, .7]] instead of [[.5, .5, .0], [.0, .3, .7]]
+        np.testing.assert_array_almost_equal(m.output_probabilities, np.array([[.5, .5, .1], [.1, .3, .7]]), decimal=2)
```
Comment (brookehus) [inline tests/markov/hmm/test_output_models.py:85]: is this what will happen in "real" cases too? is this what we want?

File: sktime/util.py [hunk 6]
```diff
@@ -266,6 +266,91 @@ def call_member(obj, f, *args, **kwargs):
     return method
 
 
+class QuantityStatistics(object):
```
Comment (brookehus) [inline sktime/util.py:269]: does this need to be unit tested or meh

**Comment by brookehus** [comment]
I gave some comments which are mostly high level/functional. I didn't see any possible bugs but I don't think I should 100% be trusted either *especially* for the C code. If there is something specific you want me to look at more deeply please let me know!

File: tests/markov/hmm/test_output_models.py (hunk 5 above)
Comment (clonker) [reply tests/markov/hmm/test_output_models.py:85]: I was actually wondering the same thing, this is replicating the bhmm behavior. It does happen if some states are not visited in a certain trajectory. One has to make sure afterwards that the output probabilities are still row-stochastic (0.5 + 0.5 + 0.1 is clearly not 1). I believe this is built in because the discrete output model in a HMM samples from the dirichlet distribution which can't be sampled if there are no counts in a state so they probabilities are just left as-is.

File: sktime/util.py (hunk 6 above)
Comment (clonker) [reply sktime/util.py:269]: optimally it would be unit tested, right now it just appears some other tests so one could say it is integration tested at least :)

File: sktime/markov/msm/markov_state_model.py (hunk 4 above)
Comment (clonker) [reply sktime/markov/msm/markov_state_model.py:229]: it _should_ not happen but it did happen during one of my tests so i thought i'd add the check

File: sktime/markov/hmm/output_model.py (hunk 3 above)
Comment (clonker) [reply sktime/markov/hmm/output_model.py:302]: old code, can be removed

File: sktime/markov/hmm/_bindings/include/OutputModelUtils.h (hunk 2 above)
Comment (clonker) [reply sktime/markov/hmm/_bindings/include/OutputModelUtils.h:52]: forgot to delete

File: sktime/base.py (hunk 1 above)
Comment (clonker) [reply sktime/base.py:20]: yes, it is just to get the signature of the method right

Comment (brookehus) [reply sktime/base.py:20]: got it i think it was just a little confusing that `deep=True` somewhere down below

File: sktime/markov/hmm/output_model.py (hunk 3 above)
Comment (brookehus) [reply sktime/markov/hmm/output_model.py:302]: lots of old code i only marked the first instance!

File: sktime/markov/msm/markov_state_model.py (hunk 4 above)
Comment (brookehus) [reply sktime/markov/msm/markov_state_model.py:229]: how do you feel about raising a warning or something?

File: tests/markov/hmm/test_output_models.py (hunk 5 above)
Comment (brookehus) [reply tests/markov/hmm/test_output_models.py:85]: so what is the "correct" fix? normalizing [0.5, 0.5, 0.1] on account of some kind of prior model? or knowing that the 0.1 is "meaningless" or preserved from the original/unsampled? would it make sense to give a warning or printout? of like certain states are unsampled? or return an object/perhaps mask somehow indicating unsampled states?

File: sktime/util.py (hunk 6 above)
Comment (brookehus) [reply sktime/util.py:269]: diplomatic, i like it

File: sktime/base.py (hunk 1 above)
Comment (clonker) [reply sktime/base.py:20]: i agree, this is the sklearn implementation though. I guess if you want to set parameters (which is where `deep=True` comes up) then you want to set them all the way

File: sktime/util.py (hunk 6 above)
Comment (clonker) [reply sktime/util.py:269]: unit tested it now anyways

File: sktime/base.py (hunk 1 above)
Comment (brookehus) [reply sktime/base.py:20]: got it. you could add a comment to clarify but obv not essential

File: sktime/util.py (hunk 6 above)
Comment (brookehus) [reply sktime/util.py:269]: woot

File: sktime/markov/msm/markov_state_model.py (hunk 4 above)
Comment (clonker) [reply sktime/markov/msm/markov_state_model.py:229]: actually ill roll back the check, numpy itself spits out a warning if a division by 0 is attempted

File: tests/markov/hmm/test_output_models.py (hunk 5 above)
Comment (clonker) [reply tests/markov/hmm/test_output_models.py:85]: leaving it as it is for now, lets discuss later
